content/chapters/compute/lab/README.md

```diff
@@ -870,13 +870,269 @@ This is because, upon creating a new thread, its TLS is initialised.
 
 ## Scheduling
 
-- TODO: find / make minimalistic library (lists, some data structure) thread-safe
-- https://github.com/Pithikos/C-Thread-Pool
-- https://github.com/aknooh/User-level-thread-library
-- https://github.com/bhaargav006/User-Thread-Library
 - https://github.com/kissen/threads
 - https://www.schaertl.me/posts/a-bare-bones-user-level-thread-library/
 
+Up to now we know that the OS decides which **thread** (not process) runs on each CPU core at each time.
+Now we'll learn about the component that performs this task specifically: **the scheduler**.
+
+There are thousands of threads running at any time in a modern OS.
+The job of the scheduler is to run and pause threads as well as allocate them to the CPU cores, with the following goals:
+
+- **fairness**: we do want all threads to get the same chance to run, i.e. run for about the same amount of time
+- **throughput**: we want to run as many threads to completion so as to complete as many tasks as we can
+
+To do this, the scheduler must decide, at given times, to suspend a thread, save its current state and let another one run in its place.
+This event is called a **context switch**.
+A context switch means changing the state of one thread (the replaced thread) from RUNNING to WAITING and the state of the replacement thread from READY / WAITING to RUNNING.
+
+- Quiz?
+
+### User-Level vs Kernel-Level Threads
+
+There are two types of threads.
+The threads you've used so far are **kernel-level threads (KLT)**.
+They are created and scheduled in the kernel of the OS.
+One of the most important of their features is that they offer true parallelism.
+With KLTs, we can truly run a program on all the cores of our CPU at once.
+But we must pay a price for this: scheduling them is very complex and context switches are costly (in terms of time), especially when switching threads belonging to different processes. 
+
+By contrast, **user-level threads (ULT)** are managed by the user space.
+More of the ULTs created by a program are generally mapped to the same kernel thread.
+If a process only creates ULTs, then they will all be mapped to the single, main kernel thread of the process.
+So if we cannot run code in parallel with ULTs, then why use them?
+Well, for I/O-intensive programs (those that do lots of network calls or file operations like reads and writes - web servers are a good example), threads are expected to perform lots of blocking calls, which causes context switches.
+In such cases, user-level threads may be useful as context switches bring less overhead between user-level threads.
```

Comment:
I'm not sure if this is correct? If a user-level thread does a blocking io call, none of the other user-level threads can run.

Reply:
Yeah, you're right. I changed the use case to something more generic: "more context switches => ULTs are better".